Build failure on Musl: type of 'readlink' is unknown

I'm trying to build MangoHud for Alpine Linux which uses Musl rather than glibc, and it seems it fails to build because of it.

```
ninja: job failed: g++ -Isrc/25a6634@@MangoHud@sha -Isrc -I../src -I../include -I../modules/ImGui/src -I. -I/usr/include/dbus-1.0 -I/usr/lib/dbus-1.0/include -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -std=c++14 -Werror=return-type -fno-math-errno -fno-trapping-math -Wno-non-virtual-dtor -Wno-missing-field-initializers -Wno-format-truncation -Os -fomit-frame-pointer -Os -fomit-frame-pointer -fPIC -pthread -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS '-DPACKAGE_VERSION="v0.4.0"' -DNDEBUG -D_GNU_SOURCE -DHAVE_PTHREAD -DUSE_GCC_ATOMIC_BUILTINS -DHAVE_TIMESPEC_GET -DHAVE___BUILTIN_BSWAP32 -DHAVE___BUILTIN_BSWAP64 -DHAVE___BUILTIN_CLZ -DHAVE___BUILTIN_CLZLL -DHAVE___BUILTIN_CTZ -DHAVE___BUILTIN_EXPECT -DHAVE___BUILTIN_FFS -DHAVE___BUILTIN_FFSLL -DHAVE___BUILTIN_POPCOUNT -DHAVE___BUILTIN_POPCOUNTLL -DHAVE___BUILTIN_UNREACHABLE '-DMANGOHUD_ARCH="64bit"' -DHAVE_NVML -DHAVE_X11 -DHAVE_DBUS -fvisibility=hidden -DVK_USE_PLATFORM_XLIB_KHR -MD -MQ 'src/25a6634@@MangoHud@sha/file_utils.cpp.o' -MF 'src/25a6634@@MangoHud@sha/file_utils.cpp.o.d' -o 'src/25a6634@@MangoHud@sha/file_utils.cpp.o' -c ../src/file_utils.cpp
In file included from /usr/include/fortify/strings.h:22,
                 from /usr/include/string.h:57,
                 from /usr/include/fortify/wchar.h:30,
                 from /usr/include/c++/9.3.0/cwchar:44,
                 from /usr/include/c++/9.3.0/bits/postypes.h:40,
                 from /usr/include/c++/9.3.0/bits/char_traits.h:40,
                 from /usr/include/c++/9.3.0/string:40,
                 from ../src/file_utils.h:2,
                 from ../src/file_utils.cpp:1:
/usr/include/fortify/unistd.h:119:1: error: type of 'readlink' is unknown
  119 | _FORTIFY_FN(readlink) ssize_t readlink(const char *__p, char *__s, size_t __n)
      | ^~~~~~~~~~~
In file included from ../src/file_utils.cpp:5:
/usr/include/fortify/unistd.h: In function 'ssize_t readlink(const char*, char*, size_t)':
/usr/include/fortify/unistd.h:125:38: error: '__orig_readlink' cannot be used as a function
  125 |  return __orig_readlink(__p, __s, __n);
      |                                      ^
../src/overlay.cpp: In function 'void render_benchmark(swapchain_stats&, overlay_params&, ImVec2&, unsigned int, uint64_t)':
../src/overlay.cpp:1047:87: warning: format not a string literal and no format arguments [-Wformat-security]
 1047 |    ImGui::TextColored(ImVec4(1.0, 1.0, 1.0, alpha / params.background_alpha), finished);
      |                                                                                       ^
../src/overlay.cpp:1052:79: warning: format not a string literal and no format arguments [-Wformat-security]
 1052 |    ImGui::TextColored(ImVec4(1.0, 1.0, 1.0, alpha / params.background_alpha), duration);
      |                                                                               ^~~~~~~~
ninja: subcommand failed
```

Thanks! Sadly it doesn't work, the same error persists.

```
ninja: job failed: g++ -Isrc/25a6634@@MangoHud@sha -Isrc -I../src -I../include -I../modules/ImGui/src -I. -I/usr/include/dbus-1.0 -I/usr/lib/dbus-1.0/include -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -std=c++14 -Werror=return-type -fno-math-errno -fno-trapping-math -Wno-non-virtual-dtor -Wno-missing-field-initializers -Wno-format-truncation -Os -fomit-frame-pointer -Os -fomit-frame-pointer -fPIC -pthread -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS '-DPACKAGE_VERSION="v0.4.0"' -DNDEBUG -D_GNU_SOURCE -DHAVE_PTHREAD -DUSE_GCC_ATOMIC_BUILTINS -DHAVE_TIMESPEC_GET -DHAVE___BUILTIN_BSWAP32 -DHAVE___BUILTIN_BSWAP64 -DHAVE___BUILTIN_CLZ -DHAVE___BUILTIN_CLZLL -DHAVE___BUILTIN_CTZ -DHAVE___BUILTIN_EXPECT -DHAVE___BUILTIN_FFS -DHAVE___BUILTIN_FFSLL -DHAVE___BUILTIN_POPCOUNT -DHAVE___BUILTIN_POPCOUNTLL -DHAVE___BUILTIN_UNREACHABLE '-DMANGOHUD_ARCH="64bit"' -DHAVE_NVML -DHAVE_X11 -DHAVE_DBUS -fvisibility=hidden -DVK_USE_PLATFORM_XLIB_KHR -MD -MQ 'src/25a6634@@MangoHud@sha/file_utils.cpp.o' -MF 'src/25a6634@@MangoHud@sha/file_utils.cpp.o.d' -o 'src/25a6634@@MangoHud@sha/file_utils.cpp.o' -c ../src/file_utils.cpp
In file included from /usr/include/fortify/strings.h:22,
                 from /usr/include/string.h:57,
                 from /usr/include/fortify/wchar.h:30,
                 from /usr/include/c++/9.3.0/cwchar:44,
                 from /usr/include/c++/9.3.0/bits/postypes.h:40,
                 from /usr/include/c++/9.3.0/bits/char_traits.h:40,
                 from /usr/include/c++/9.3.0/string:40,
                 from ../src/file_utils.h:2,
                 from ../src/file_utils.cpp:1:
/usr/include/fortify/unistd.h:119:1: error: type of 'readlink' is unknown
  119 | _FORTIFY_FN(readlink) ssize_t readlink(const char *__p, char *__s, size_t __n)
      | ^~~~~~~~~~~
In file included from ../src/file_utils.cpp:5:
/usr/include/fortify/unistd.h: In function 'ssize_t readlink(const char*, char*, size_t)':
/usr/include/fortify/unistd.h:125:38: error: '__orig_readlink' cannot be used as a function
  125 |  return __orig_readlink(__p, __s, __n);
      |                                      ^
ninja: subcommand failed
```

--- a/pkgs/tools/graphics/mangohud/combined.nix
+++ b/pkgs/tools/graphics/mangohud/combined.nix
@@ -1,0 +1,14 @@
+{ stdenv, pkgs, lib
+, libXNVCtrl
+}:
+let
+  mangohud_64 = pkgs.callPackage ./default.nix { inherit libXNVCtrl; };
+  mangohud_32 = pkgs.pkgsi686Linux.callPackage ./default.nix { inherit libXNVCtrl; };
+in
+pkgs.buildEnv rec {
+  name = "mangohud-${mangohud_64.version}";
+
+  paths = [ mangohud_32 ] ++ lib.optionals stdenv.is64bit [ mangohud_64 ];
+
+  meta = mangohud_64.meta;
+}

--- a/pkgs/tools/graphics/mangohud/default.nix
+++ b/pkgs/tools/graphics/mangohud/default.nix
@@ -1,0 +1,79 @@
+{ stdenv
+, lib
+, fetchFromGitHub
+, fetchpatch
+, meson
+, ninja
+, pkg-config
+, python3Packages
+, dbus
+, glslang
+, libglvnd
+, libXNVCtrl
+, mesa
+, vulkan-headers
+, vulkan-loader
+, xorg
+}:
+
+
+stdenv.mkDerivation rec {
+  pname = "mangohud${lib.optionalString stdenv.is32bit "_32"}";
+  version = "0.4.1";
+
+  src = fetchFromGitHub {
+    owner = "flightlessmango";
+    repo = "MangoHud";
+    rev = "v${version}";
+    sha256 = "04v2ps8n15ph2smjgnssax5hq88bszw2kbcff37cnd5c8yysvfi6";
+    fetchSubmodules = true;
+  };
+
+  patches = [
+    (fetchpatch {
+      # FIXME obsolete in >=0.5.0
+      url = "https://patch-diff.githubusercontent.com/raw/flightlessmango/MangoHud/pull/208.patch";
+      sha256 = "1i6x6sr4az1zv0p6vpw99n947c7awgbbv087fghjlczhry676nmh";
+    })
+  ];
+
+  mesonFlags = [
+    "-Dappend_libdir_mangohud=false"
+    "-Duse_system_vulkan=enabled"
+    "--libdir=lib${lib.optionalString stdenv.is32bit "32"}"
+  ];
+
+  buildInputs = [
+    dbus
+    glslang
+    libglvnd
+    libXNVCtrl
+    mesa
+    python3Packages.Mako
+    vulkan-headers
+    vulkan-loader
+    xorg.libX11
+  ];
+
+  nativeBuildInputs = [
+    meson
+    ninja
+    pkg-config
+    python3Packages.Mako
+    python3Packages.python
+  ];
+
+  preConfigure = ''
+    mkdir -p "$out/share/vulkan/"
+    ln -sf "${vulkan-headers}/share/vulkan/registry/" $out/share/vulkan/
+    ln -sf "${vulkan-headers}/include" $out
+  '';
+
+  meta = with lib; {
+    description = "A Vulkan and OpenGL overlay for monitoring FPS, temperatures, CPU/GPU load and more";
+    homepage = "https://github.com/flightlessmango/MangoHud";
+    platforms = platforms.linux;
+    license = licenses.mit;
+    maintainers = with maintainers; [ zeratax ];
+  };
+}

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -6532,6 +6532,10 @@ in
 
   mandoc = callPackage ../tools/misc/mandoc { };
 
+  mangohud = callPackage ../tools/graphics/mangohud/combined.nix {
+    libXNVCtrl = linuxPackages.nvidia_x11.settings.libXNVCtrl;
+  };
+
   manix = callPackage ../tools/nix/manix {
     inherit (darwin.apple_sdk.frameworks) Security;
   };

